src/DomainGuest.sol

```diff
@@ -0,0 +1,365 @@
+        if (what == "dust") dust = data;
+        else revert("DomainGuest/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    // --- Canonical DAI Support ---
+
+    /// @notice Mint DAI and send to user
+    /// @param to The address to send the DAI to on the local domain
+    /// @param amount The amount of DAI to send [WAD]
+    function _deposit(address to, uint256 amount) internal {
+        vat.swell(address(this), _int256(amount * RAY));
+        daiJoin.exit(to, amount);
+
+        emit Deposit(to, amount);
+    }
+
+    /// @notice Withdraw DAI by burning local canonical DAI
+    /// @param to The address to send the DAI to on the remote domain
+    /// @param amount The amount of DAI to withdraw [WAD]
+    function _withdraw(address to, uint256 amount) internal {
+        require(dai.transferFrom(msg.sender, address(this), amount), "DomainGuest/transfer-failed");
+        daiJoin.join(address(this), amount);
+        vat.swell(address(this), -_int256(amount * RAY));
+
+        emit Withdraw(msg.sender, to, amount);
+    }
+
+    // --- MCD Support ---
+
+    /// @notice Record changes in line and grain and update dss global debt ceiling if necessary
+    /// @param _lid Local ordering id
+    /// @param wad The new debt ceiling [WAD]
+    function _lift(uint256 _lid, uint256 wad) internal ordered(_lid) {
+        require(live == 1, "DomainGuest/not-live");
+
+        uint256 lastLine = vat.Line();
+        uint256 rad = wad * RAY;
+        if (rad > lastLine) grain = wad;
+        vat.file("Line", rad);
+
+        emit Lift(wad);
+    }
+
+    /// @notice Will release remote DAI from the escrow when it is safe to do so
+    /// @dev    Should be run by keeper on a regular schedule.
+    function _release() internal returns (uint256 _rid, uint256 _burned) {
+        require(live == 1, "DomainGuest/not-live");
+
+        uint256 limit = _max(vat.Line() / RAY, _divup(vat.debt(), RAY));
+        require(grain >= limit + dust / RAY, "DomainGuest/dust");
+        uint256 burned = grain - limit;
+        grain = limit;
+
+        _rid = rid++;
+        _burned = burned;
+
+        emit Release(burned);
+    }
+
+    /// @notice Push surplus (or deficit) to the host dss
+    /// @dev Should be run by keeper on a regular schedule
+    function _push() internal returns (uint256 _rid, int256 _surplus) {
+        require(live == 1, "DomainGuest/not-live");
+
+        _rid = rid++;
+
+        uint256 _dai = vat.dai(address(this));
+        uint256 _sin = vat.sin(address(this));
+        if (_dai >= _sin + dust) {
+            // We have a surplus
+            if (_sin > 0) vat.heal(_sin);
+
+            uint256 wad = (_dai - _sin) / RAY;    // Round against this contract for surplus
+
+            // Burn the DAI and unload on the other side
+            vat.swell(address(this), -_int256(wad * RAY));
+            _surplus = _int256(wad);
+
+            emit Push(int256(wad));
+        } else if (_sin >= _dai + dust) {
```

Comment:
Not sure if I'm missing something, but couldn't this message be produced a ton of times provoking in L1 an acumulation of a huge unexisting `sin`?
Still only governance can `rectify` but feels that this should be prevented somehow. Maybe keeping a storage `deficit` in the guest, which will be cleaned in rectify.

Reply:
If this is actually correct (seems Oli agrees), I can work on a PR for a solution and I think I'd pursue on separating accounting surplus and deficit in two separated functions.

Reply:
Yes looks like it's indeed missing some state change in `DomainGuest._push()` in the deficit case.

Reply:
Ah good catch yes this needs to monitor the pending deficit on the guest side.

Reply:
Ok good, I can work on the PR after finishing the documentation. Will probably split it as mentioned and we can discussed in that PR if looks good or get back to unified function.